### ISSUE TITLE
Data Mapping for Different OSes

### ISSUE THREAD

**Issue by ColonelPanicks** [alces-software/inventoryware#7]
To keep input templates platform-agnostic requires a way of sanitising the data read in from the various command outputs based on the OS.

For example, on EL6 systems the value for "CPU Model" in lshw would be under `['lshw']['cpuX']['version']` whereas on EL7 it's `['lshw']['cpuX']['model']`. Instead of changing the templates to handle this difference would it be possible to map a key to a different value depending on OS. This will allow putting something arbitrary like `hash['map']['cpu_model']` into a template.

**Comment by ColonelPanicks** [comment]
I'm afraid I can't off of the top of my head but in more general terms it can allow for prettifying data that perhaps doesn't differ between OSes. Like in your original templates where the internals of Underware were setting `hash['BIOS Version']` to wherever that value was tucked away, being able to create these bonds with a more content-like approach.

**Comment by DavidMarchant** [comment]
I can't off the top of my head think of a generalized way to do this - the information on in what way the field is differently stored will itself need to be stored somewhere  - as it stands, as all "objects" exist as a hash, we could just use a method per field for the time being. (possibly until I have a better idea of different examples, in which case I can work on a general solution)

**Comment by ColonelPanicks** [comment]
Ahh yeah I figured it'd get tricky when the CPU will be a hash of CPUs. The concern with it being in util.rb is that it becomes a function that's then maintained as part of inventoryware and isn't as straightforward to tweak or add new OSes.

What if mappings have the current object passed through inherently or if one is specified? So if I were to do something like:
```
Model: <%= mappings['cpu_model'](cpu) %>
```

And in a content file I can add the mapping for the different OSes:
```
el6:
  cpu_model: hash['model']
el7:
  cpu_model: hash['version']
```

This then generalises the lookup method and may leave it a bit more flexible in terms of content.

**Comment by DavidMarchant** [comment]
Ah sure! Cheers, I'll get right on with implementing that tomorrow

**Comment by ColonelPanicks** [comment]
After review and discussion with @DavidMarchant and @mjtko it has been decided that mapping will be handled with specific ruby files that will be considered `plugins` which utilise functions to build up OpenStruct objects that the templates can then use.

**Comment by DavidMarchant** [comment]
The above specified solution - using 'plugin' ruby files - has been implemented with #20. As such I will now close this issue.
